### DISCUSSION TITLE
CitriNet finetuning not going as expected

### DISCUSSION THREAD

**Discussion by francescodaq** [NVIDIA-NeMo/NeMo#3106] [Q&A]
Hello,

we are trying to obtain a good ASR model for italian language, selecting CitriNet as model.
As first step we started from the Nvidia provided pre-trained english model `STT En Citrinet 1024` and followed this tutorials (https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/ASR_with_Subword_Tokenization.ipynb and especially https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/ASR_CTC_Language_Finetuning.ipynb) for transfer learning to italian language.

We performed a training session of 200 epochs on Mozilla CommonVoice dataset (6.1) obtaining a wer of 20%.
These were the parameters we used for optimizer:

```
optim:
  name: novograd
  lr: 0.025
  betas:
  - 0.8
  - 0.25
  weight_decay: 0.001
  sched:
    name: CosineAnnealing
    warmup_steps: null
    warmup_ratio: 0.1
    min_lr: 1.0e-09
    last_epoch: -1
```
  
and for spec augmentation:

```
spec_augment:
  _target_: nemo.collections.asr.modules.SpectrogramAugmentation
  freq_masks: 2
  time_masks: 10
  freq_width: 25
  time_width: 0.05
```

Batch sizes for both train and validation were set to 16.

The tokenizer (1024 tokens) was built on text corpus of train+dev+test dataset from common voice and another dataset we had (let's say dataset B).
Both encoder and decoder were unfrozen. Weights from decoder of the pretrained model were kept and restored after changing vocabulaty, as explained in the tutorial notebook.
Training was performed on a single GPU.

We are aiming to improve wer for italian language, so we set up a script for further finetune the model performing another training session on an updated version of italian common voice dataset (7.0).
In order to speedup the training process we moved from a single to four GPU machine.

These are the step we performed before starting the new training:

 - rebuild the tokenizer using as corpus text from train+test+dev of common voice 7.0 dataset, dataset B and dataset C, keeping number of token 1024 
 - restore the model trained in our previous session
 - update the vocab and restoring decoeder weights 

These are the parameters used for optimizer (we adjusted learning rate due to the increment of GPU number by multiplying it of a factor sqrt(4)=2)
```

optim:
	name: novograd
	lr: 0.05
	betas:
	- 0.8
	- 0.25
	weight_decay: 0.001
	sched:
		name: CosineAnnealing
		warmup_steps: null
		warmup_ratio: 0.1
		min_lr: 1.0e-09
		last_epoch: -1
```
		
	
We kept unchanged values for spec augmentation:
```

spec_augment:
  _target_: nemo.collections.asr.modules.SpectrogramAugmentation
  freq_masks: 2
  time_masks: 10
  freq_width: 25
  time_width: 0.05  
```

Batch size for both train and validation is set to 12.

The following picture is about how first and second training ended / are going:

![Untitled](https://user-images.githubusercontent.com/87667235/139580197-16d3f2fd-323f-406f-9664-ad251806b23a.png)

Orange line is about the first session training, the green one is about the second training.

This is the python script we are using for second training session:

[FineTuneFromPretrained-withTokenizer.py.txt](https://github.com/NVIDIA/NeMo/files/7448626/FineTuneFromPretrained-withTokenizer.py.txt)

Training is running but it seems that after warmup period WERs and loss are stuck and keeping a more or less constant value, it seems model is not learning.
Do you think it is worth keep it going or stop the training?
Is there something wrong in how we are performing the second fine tuning session? Do you have any advice about how to change parameters?

Thanks,
Francesco

**Comment by titu1994** [comment]
This is very detailed, thank you for the info ! Your approach is very good, and I see no mistakes in the code. 

I would suggest two things - use warmup steps of explicitly 10000 (right now it's 10% of training steps ~ 100k steps).

The learning rate is far too high. Novograd is very stable to learning rate, there's no need to scale it via sqrt scaling especially for finetuning on same language. We use 0.05 LR for 32 GPUs to 512 gpus. For small number of GPUs, 0.025 or 0.01 may be better.  

Since the first iteration model was trained on Italian, you can actually reduce learning rate to 0.01 instead. Also, to perhaps speed up convergence, as long as the base character set has remained the same, you could try using the tokekizer built for the first model itself. Of course it is perfectly valid to rebuild the tokenizer, it will take slightly longer to converge. 

I see that the wer jumps midway between warmup, telltale sign that the LR was too high causing pretrainrd weights being updated too strongly. You'll need to adjust the LR such that the spike doesn't occur - in the range of 0.01 to 0.001

**Reply by francescodaq** [reply]
Thank you so much for the answer!

We decided to stop the training session and start a new one following your advices.

Changes:
 - lowered `lr `to 0.009
 - Due to `lr `reduction we decreased `weight decay` too (from 0.001 to 0.0005).
 - set `warmup steps` explicitly (10000)
 - old `tokenizer `has been kept unchanged

These are new optimization parameters:

```
optim:
  name: novograd
  lr: 0.009
  betas:
  - 0.8
  - 0.25
  weight_decay: 0.0005
  sched:
    name: CosineAnnealing
    warmup_steps: 10000
    warmup_ratio: null
    min_lr: 1.0e-09
    last_epoch: -1
```

We will keep updated about new training session results.

Thanks!
